Move crypto functions into their own file for easy replacement

dwmw2 opened the pull request:
On top of the `async` branch from PR #8 this moves the crypto (and crc32) functionality to methods of the base tuyaAPI class. This is the final piece of the puzzle before I'm able to build an ESPHome component which uses tuyaAPI:

https://github.com/dwmw2/esphome/commit/tuya_tcp

In that, I use the `tuyaAPI*` files directly (currently symlinks; I'll work on that).

It provides a replacement for tuyaTCP which is just an empty class (with a comment saying that tuyaTCP would ideally be a parent class which has a tuyaAPI member, but I can live with this for now).

dwmw2 committed:
Promote NegotiateSession to first-class tuyaAPI method

Add a NegotiateSession() method. For the protocols before 3.4 which don't
do actual session negotiation, this just sets the encryption key.

For the 3.4 protocol it does the full negotiation on the wire as it did
before, now under the explicit control of the user instead of implicitly
blocking in the first BuildTuyaMessage() call.

This change also removes the encryption_key parameter from BuildTuyaMessage()
and DecodeTuyaMessage() since the key is now stored in a member variable.

The examples now call NegotiateSession(device_key) after ConnectToDevice().

dwmw2 committed:
Move NegotiateSession() to base tuyaAPI class

This makes negotiation in the actual protocols properly non-blocking,
with BuildSessionMessage() and DecodeSessionMessage() methods. All
actual TCP communication is driven outside the protocol code.

NegotiateSession() is now just a simple loop which uses those functions
along with the blocking TCP send() and receive() functions — and like
those TCP functions, it's *optional*. A caller who wants properly
non-blocking operation can just call the protocols directly instead.

There are default BuildSessionMessage()/DecodeSessionMessage() methods
for the older protocols which don't do negotiation on the wire, so
the call to NegotiateSession() only sets the encryption key, as before.

dwmw2 committed:
Add crypto abstraction layer for alternative implementations

Add virtual methods for crypto operations to allow alternative implementations:
- aes_128_ecb_encrypt/decrypt: AES-128-ECB encryption/decryption
- hmac_sha256: HMAC-SHA256 computation
- md5_hash: MD5 hash computation
- crc32_compute: CRC32 checksum
- random_bytes: Cryptographically secure random bytes

Convert all protocol files to use these abstraction methods instead of
calling OpenSSL/zlib directly. Default implementation using OpenSSL and
zlib is in tuyaAPI-crypto.cpp, which can be easily replaced with mbedtls
or other crypto libraries.

The decrypt function handles padding errors gracefully to match the
original behavior where EVP return values weren't checked.

Not all protocols use all of these functions, but if only one protocol
is used then the unused functions should naturally drop out of the final
linked executable. And they're tiny anyway.

gordonb3 commented:
You're absolutely on to something here, but we should use this as re-enterable code rather than replicate it.

dwmw2 commented:
Hm, what exactly do you mean? There are a bunch of crypto operations the protocols use, and rather than having them open-code calls to OpenSSL we just pull those common duplicated sections of code out to library functions. It used to be replicated across the protocols, and now it's no longer replicated. Each is moved out to a single function that all the protocols can call. And even better, different environments can replace the implementation of those functions.

dwmw2 commented:
Updated to stop them being *virtual* functions. If they're virtual, the compiler emits them into the vtable of every `tuyaAPI3x` class, and they have to be present. If they're just standard functions, then any optimising setup using LTO or `--gc-sections` can drop the functions entirely if they aren't used.

dwmw2 commented:
Verified in the ESPHome build with `--gc-sections` that *all* the functions related to unused protocols just fall away when not used, so if I don't configure a v3.5 device the GCM crypto functions aren't even present in the executable.

dwmw2 committed:
Refactor tuyaAPI34 to eliminate code duplication

Extract common message-building code into BuildMessage34() helper function.
Both BuildTuyaMessage() and BuildSessionMessage() now use this helper,
eliminating ~40 lines of duplicated header/encryption/HMAC/trailer code.
